rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py

```diff
@@ -615,8 +638,33 @@ def filter_waypoints(self, wps: list, threshold=1.0):
 
         return (first, second)
 
+    def set_action_execution(self, action_execution):
+        if self.action_execution is None:
```

Comment:
We should be able to overwrite the `action_execution` and not only when its `None`. So let's remove this `if` check.

Also, why are we re-initializing all these `self.` parameters? We should just lock and set `self.action_execution`?

Reply:
Hmm overwriting `action_execution` isn't under the `if` condition. The `if` check ensures that we are setting a new `action_execution`, so that we can initialize `self.action_waypoint_index` (new `self.` parameter added) used for updating robot position during the action. Without it the purple marker doesn't move. Not sure if there's a better way to do this. I can move the lock out of the `if` check where action execution is set.

Reply:
Ah whoops my bad, it's not in the `if` scope. Another reason to hate python indentation 🤦🏼‍♂️ 

But you should be setting these parameters inside the `_action_executor()` [callback](https://github.com/xiyuoh/rmf_demos/blob/96415381503863d71a4a7f809f60c265c87d7fad/rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py#L471). The values you set here will get overwritten in the next update timer iteration. Also, why are we assigning`self.action_waypoint_index` with `self.on_waypoint`? When the robot starts performing the action, it will be trying to merge into a different waypoint at the end of the action and high chance that won't be the `self.on_waypoint` that was initialized. Just noticed this error. So when you call [update_off_grid()](https://github.com/xiyuoh/rmf_demos/blob/96415381503863d71a4a7f809f60c265c87d7fad/rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py#L504-L506), you're giving RMF the current position of the robot and telling it that the robot is trying to merge back into `self.action_waypoint_index` which is not set correctly. You should set the value correctly inside the `_action_executor()` callback. And the waypoint should come from the [description](https://github.com/xiyuoh/rmf_demos/blob/96415381503863d71a4a7f809f60c265c87d7fad/rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py#L472) field. If this object is not being populated now (because the task submission description does not have it, then you should modify that submission script to have that information. But if that information is not provided, the action waypoint should be set to whatever the robot's on_waypoint is at the time the action begins execution. Again this can be done by assigning within `_action_executor()`

Reply:
Oh okay i see! I misunderstood the purpose of passing the waypoint index to `update_off_grid()`. Just to clarify, you are saying that in case description is not provided, we should set the `self.action_waypoint_index` to the starting waypoint when RMF relinquish control, i.e. what is currently written? 

Do you think we can set `self.action_waypoint_index` to the nearest waypoint on the map instead when `complete_robot_action` is called? Something like this:

```
def complete_robot_action(self):
        if self.action_execution is None:
            return
        if not self.action_description:
            nearest_dist = np.inf
            nearest_wp = None
            for i in range(self.graph.num_waypoints):
                wp = self.graph.get_waypoint(i)
                dist_to_wp = self.dist(self.position, wp.location)
                if dist_to_wp < nearest_dist:
                    nearest_dist = dist_to_wp
                    nearest_wp = wp
            self.action_waypoint_index = nearest_wp.index
            self.update_state()
        with self._lock:
            self.action_execution.finished()
            self.action_execution = None
            self.node.get_logger().info(f"Robot {self.name} has completed the"
                                        f" action it was performing")
```